src/app/AlwaysOn.Shared/TelemetryExtensions/AppInsightsRequestHandler.cs

```diff
@@ -0,0 +1,73 @@
+﻿using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.Azure.Cosmos;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace AlwaysOn.Shared.TelemetryExtensions
+{
+    public class AppInsightsCosmosRequestHandler : RequestHandler
+    {
+        public AppInsightsCosmosRequestHandler(TelemetryClient telemetryClient)
+        {
+            _telemetryClient = telemetryClient;
+        }
+
+        private readonly TelemetryClient _telemetryClient;
+
+        public override async Task<ResponseMessage> SendAsync(RequestMessage request, CancellationToken cancellationToken)
+        {
+            var title = request.Properties.ContainsKey("Operation") ? request.Properties["Operation"].ToString() : $"{request.Method} {request.RequestUri.OriginalString}";
+            var dbClientEndpoint = request.Properties.ContainsKey("DbClientEndpoint") ? request.Properties["DbClientEndpoint"].ToString() : null;
+
+            using var dependency = _telemetryClient.StartOperation<DependencyTelemetry>(title);
+
+            //
+            // Comparison with the original telemetry object:
+            //
+            //var telemetry = new DependencyTelemetry()
+            //{
+            //    OK - Type = AppInsightsDependencyType,
+            //    OK - Data = $"ObjectId={objectId}, Partitionkey={partitionKey}",
+            //    OK - Name = $"Delete {typeof(T).Name}",
+            //    OK - Timestamp = startTime,
+            //    OK - Duration = diagnostics != null ? diagnostics.GetClientElapsedTime() : overallDuration,
+            //    OK - Target = diagnostics != null ? diagnostics.GetContactedRegions().FirstOrDefault().uri?.Host : _dbClient.Endpoint.Host,
+            //    REMOVED - Success = success
+            //};
+            //
+            //if (response != null)
+            //    OK - telemetry.Metrics.Add("CosmosDbRequestUnits", response.RequestCharge);
+
+            var response = await base.SendAsync(request, cancellationToken);
+
+            var success = response.IsSuccessStatusCode;
+            // Application-specific special handling in order to have cleaner Application Insights error reporting. Client will still get proper error response.
+            // There are cases which we don't consider database errors and would create noise in monitoring:
+            //  - Considering NotFound as successful call to Cosmos DB, looking for a non-existent object.
+            //  - The same with Conflict, which is handled the application code (ignoring the request, because it already exists).
+            if (response.StatusCode == System.Net.HttpStatusCode.NotFound || response.StatusCode == System.Net.HttpStatusCode.Conflict)
+                success = true;
+
+            dependency.Telemetry.Type = "Azure DocumentDB";
```

Comment:
I wonder if we should add a try-catch around the tracing. just to be very sure. Otherwise it could happen that the request itself worked, but an exception somewhere down here would be thrown up and the caller would assume the operation to cosmos failed?! not sure if something still can go wrong, but I have seen NREs with the Diagnostics before...

Reply:
Good point, we probably want the response to be returned in any case :D

Reply:
Updated

Reply:
great